### PR TITLE
Prevent AD0001 on missing analyzer file

### PR DESCRIPTION
**Pull request by mayerj**

### CODE REVIEW

**Review by mavasani — COMMENTED** [review]
Do we have an issue tracking this? Are you able to write a regression test?

**Comment by mayerj** [comment]
> Do we have an issue tracking this? Are you able to write a regression test?

This is issue [5061](https://github.com/dotnet/roslyn-analyzers/issues/5061)

I tried writing a test, but have run into some issues.  The "AdditionalFiles" in the unit tests don't support a null value for the text (just not present, or string.empty).  However, the issue is reproducible.

https://github.com/dotnet/roslyn-analyzers/blob/a0522afd52b25ad211298e308fd5bbc7734865e0/src/PublicApiAnalyzers/UnitTests/DeclarePublicApiAnalyzerTests.cs#L62

I am still attempting to write the test differently.

**Comment by mayerj** [comment]
@mavasani I'm not sure it's straightforward to write a test.  I need to generate an AdditionalText which returns null from GetText, but it seems like the unit test framework seems to deal in SourceText objects themselves.  Any hints?

**Review by Youssef1313 — COMMENTED** [review]
Can you add a test? Specifically for https://github.com/dotnet/roslyn-analyzers/issues/5061#issuecomment-825695032.

---

> I'm not sure it's straightforward to write a test. I need to generate an AdditionalText which returns null from GetText, but it seems like the unit test framework seems to deal in SourceText objects themselves. Any hints?

I'm not 100% sure, but I think `build_property.AdditionalFiles` may be helpful to represent an `<AdditionalFiles>` in csproj that doesn't actually exist on disk.

It may look something like:


```csharp

await new VerifyCS.Test
{
    TestState =
    {
        AdditionalFiles = {  }, // Empty that they are not existing on disk.
        AnalyzerConfigFiles = { ("/.globalconfig", @"
is_global = true

build_property.AdditionalFiles = PublicAPI.Shipped.txt
"),
    }
}
```

**Comment by mayerj** [comment]
> Can you add a test? Specifically for [#5061 (comment)](https://github.com/dotnet/roslyn-analyzers/issues/5061#issuecomment-825695032).
> 
> > I'm not sure it's straightforward to write a test. I need to generate an AdditionalText which returns null from GetText, but it seems like the unit test framework seems to deal in SourceText objects themselves. Any hints?
> 
> I'm not 100% sure, but I think `build_property.AdditionalFiles` may be helpful to represent an `<AdditionalFiles>` in csproj that doesn't actually exist on disk.
> 
> It may look something like:
> 
> ```cs
> await new VerifyCS.Test
> {
>     TestState =
>     {
>         AdditionalFiles = {  }, // Empty that they are not existing on disk.
>         AnalyzerConfigFiles = { ("/.globalconfig", @"
> is_global = true
> 
> build_property.AdditionalFiles = PublicAPI.Shipped.txt
> "),
>     }
> }
> ```

I don't think so?  I thought that that functionality was to move the CSPROJ data into the analyzers, but I need the file provided as an AdditionalFile in analyzerOptions.AdditionalFiles.  The analyzer isn't looking for additional files in the editorconfig data (and at least my brief testing didn't translate the snippet above into an analyzerOptions.AdditionalFiles entry, is it supposed to?

**Comment by Youssef1313** [comment]
I'll defer to @sharwell if you tried that and it didn't work. My understanding is that `build_property.*` translates as if the property is present in csproj.

**Comment by sharwell** [comment]
Writing a test requires a workspace implementation that believes a file exists only to find that it doesn't actually exist when you try to read it. It might be possible through a solution transform, but I wouldn't count on it.

**Comment by mayerj** [comment]
> Writing a test requires a workspace implementation that believes a file exists only to find that it doesn't actually exist when you try to read it. It might be possible through a solution transform, but I wouldn't count on it.

I'm certainly willing to do it, but I will need some pointers if that is the desire.  I did play with the solution transform, and it doesn't seem like there is a way of getting at the AdditionalFiles list to add directly to it.  I can only add SourceText objects, which are checked for null prior to the test running.

**Comment by sharwell** [comment]
If you wanted to try it, you could do something like this:

```csharp
await new VerifyCS.Test
{
  // other properties ...
  SolutionTransforms =
  {
    (solution, projectId) =>
    {
      var id = DocumentId.CreateNewId(projectId, "non-existent file");
      var fileName = Path.GetRandomFileName();
      var filePath = Path.Combine(Path.GetTempPath(), fileName);
      return solution.AddAdditionalDocument(DocumentInfo.Create(id, fileName, loader: new FileTextLoader(filePath, Encoding.UTF8), filePath: filePath));
    },
  },
}.RunAsync();
```

**Comment by mayerj** [comment]
> ```cs
> await new VerifyCS.Test
> {
>   // other properties ...
>   SolutionTransforms =
>   {
>     (solution, projectId) =>
>     {
>       var id = DocumentId.CreateNewId(projectId, debugName: "non-existent file");
>       var fileName = Path.GetRandomFileName();
>       var filePath = Path.Combine(Path.GetTempPath(), fileName);
>       return solution.AddAdditionalDocument(DocumentInfo.Create(id, fileName, loader: new FileTextLoader(filePath, Encoding.UTF8), filePath: filePath));
>     },
>   },
> }.RunAsync();
> ```

I thought I had it, but this doesn't throw the NullReferenceException.  It seems like in the test this results in an "AdditionalTextWithState" object to be returned in AdditionalFiles.  However, in VS2019/the compiler, it returns an "AdditionalTextFile".

The AdditionalTextWithState has some additional logic that returns a "failed" text if it can't  read the document.

https://github.com/dotnet/roslyn/blob/315c2e149ba7889b0937d872274c33fcbfe9af5f/src/Workspaces/Core/Portable/Workspace/Solution/TextLoader.cs#L60

The AdditionalTextFile does seem to hint that it might return null.

https://github.com/dotnet/roslyn/blob/3aa88b7957cbb4dedcc03f2fa58f79bf9a68538a/src/Compilers/Core/Portable/AdditionalTextFile.cs#L53

**Comment by sharwell** [comment]
💭 It almost seems like this should be fixed by updating the compiler to avoid exposing these null values to the analyzers.

**Comment by mayerj** [comment]
> I don't see a way to bypass this difference without private reflection.
> 
> 💭 It almost seems like this should be fixed by updating the compiler to avoid exposing these null values to the analyzers.

Agreed, though it does seem intentional that it can return null.

https://github.com/dotnet/roslyn/blob/aa91123c53d9a84daf6053c69e88fff255c236d4/src/Compilers/Core/Portable/DiagnosticAnalyzer/AdditionalText.cs#L24

What is next for the PR?

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.cs
Comment (sharwell) [inline src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.cs:322]: ❔ Is this better than just updating `ReadApiData` to check the argument for null?

**Comment by sharwell** [comment]
@mayerj I have no objection to taking this change without tests. I wasn't really expecting them to start with, but we still gave it a solid shot. I did have question about the value of this "more complicated" approach over a simpler approach that involves a null check. If you have a reason to prefer the current form feel free to respond with an explanation and not feel obligated to make a change.

**Comment by mavasani** [comment]
Agree with @sharwell - thank you @mayerj for putting in the hard yards to create a test. We should be good to merge once @sharwell approves.

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.cs
Comment (mayerj) [reply src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.cs:322]: The reason I went with this approach was that I wanted to treat the file being configured but not available on disk the same as if it was unconfigured (IE, it would respect silentlyBailOutOnMissingApiFiles if they were both missing, but give you PublicApiFileMissing if it was just 1).  So TryGetApiText needs to return false if shippedText.GetText(...) is going to return null.  At that point I just figured there was no reason to call it twice, so I just returned the path and content.

**Review by mavasani — APPROVED** [review]

**Pull request merged**
